File: web/scripts/preprocess.py
import json
import os
import shutil
from dataclasses import dataclass
from pathlib import Path
from typing import Any, Dict, List, Tuple

import pandas as pd
import pyarrow.parquet as pq


RAW_DATA_DEFAULT = Path(__file__).resolve().parents[2] / "data" / "player_data" / "player_data"
OUT_DATA_DEFAULT = Path(__file__).resolve().parents[1] / "public" / "data"
OUT_MATCHES_DIRNAME = "matches"


MAP_CONFIG: dict[str, dict[str, float]] = {
    "AmbroseValley": {"scale": 900, "origin_x": -370, "origin_z": -473},
    "GrandRift": {"scale": 581, "origin_x": -290, "origin_z": -290},
    "Lockdown": {"scale": 1000, "origin_x": -500, "origin_z": -500},
}


def is_numeric_user_id(user_id: str) -> bool:
    return user_id.isdigit()


def decode_event(v: Any) -> str:
    if isinstance(v, (bytes, bytearray)):
        return v.decode("utf-8", errors="replace")
    return str(v)


def datetime_to_ms(dt_series: pd.Series) -> pd.Series:
    # parquet timestamps come as datetime64[ns] in pandas; convert to integer ms
    return (dt_series.astype("int64") // 1_000_000).astype("int64")


@dataclass
class PlayerBlob:
    user_id: str
    is_bot: bool
    positions: List[Tuple[int, float, float]]  # (t_ms_abs, x, z)
    events: List[Dict[str, Any]]  # {t_ms_abs, x, z, type}


@dataclass
class MatchBlob:
    match_id: str
    map_id: str
    date: str
    players: Dict[str, PlayerBlob]


def iter_player_files(raw_root: Path) -> List[Tuple[str, Path]]:
    days: list[tuple[str, Path]] = []
    for child in sorted(raw_root.iterdir()):
        if child.is_dir() and child.name.startswith("February_"):
            days.append((child.name, child))

    out: list[tuple[str, Path]] = []
    for day_name, day_path in days:
        for f in day_path.iterdir():
            if not f.is_file():
                continue
            if f.name.startswith("."):
                continue
            out.append((day_name, f))
    return out


def parse_user_id_from_filename(filename: str) -> str:
    # {user_id}_{match_id}.nakama-0
    base = filename.split(".")[0]  # drop .nakama-0 and anything after first dot
    return base.split("_")[0]


def main() -> None:
    raw_root = Path(os.environ.get("RAW_PLAYER_DATA_DIR", str(RAW_DATA_DEFAULT))).resolve()
    out_root = Path(os.environ.get("OUT_PUBLIC_DATA_DIR", str(OUT_DATA_DEFAULT))).resolve()
    out_matches = out_root / OUT_MATCHES_DIRNAME

    if not raw_root.exists():
        raise SystemExit(f"Raw data folder not found: {raw_root}")

    out_matches.mkdir(parents=True, exist_ok=True)

    # Copy minimaps into web/public/minimaps (so Next can serve them)
    minimaps_src = raw_root / "minimaps"
    minimaps_dst = out_root.parents[0] / "minimaps"
    minimaps_dst.mkdir(parents=True, exist_ok=True)
    if minimaps_src.exists():
        for img in minimaps_src.iterdir():
            if img.is_file() and img.suffix.lower() in [".png", ".jpg", ".jpeg", ".webp"]:
                shutil.copy2(img, minimaps_dst / img.name)

    files = iter_player_files(raw_root)
    matches: Dict[str, MatchBlob] = {}

    for day_name, file_path in files:
        user_id = parse_user_id_from_filename(file_path.name)
        bot = is_numeric_user_id(user_id)

        try:
            table = pq.read_table(file_path)
            df = table.to_pandas()
        except Exception:
            # skip any unreadable file
            continue

        if df.empty:
            continue

        df["event"] = df["event"].apply(decode_event)
        df["t_ms"] = datetime_to_ms(df["ts"])

        match_id = str(df["match_id"].iloc[0])
        map_id = str(df["map_id"].iloc[0])

        # sanity: keep only maps we know
        if map_id not in MAP_CONFIG:
            continue

        match = matches.get(match_id)
        if match is None:
            match = MatchBlob(match_id=match_id, map_id=map_id, date=day_name, players={})
            matches[match_id] = match

        # If a match appears in multiple day folders (unlikely), keep the first date we saw.
        if match.map_id != map_id:
            # ignore inconsistent data
            continue

        pb = match.players.get(user_id)
        if pb is None:
            pb = PlayerBlob(user_id=user_id, is_bot=bot, positions=[], events=[])
            match.players[user_id] = pb

        pos_mask = df["event"].isin(["Position", "BotPosition"])
        pos_df = df.loc[pos_mask, ["t_ms", "x", "z"]].sort_values("t_ms")
        if not pos_df.empty:
            pb.positions.extend([(int(t), float(x), float(z)) for t, x, z in pos_df.itertuples(index=False, name=None)])

        ev_df = df.loc[~pos_mask, ["t_ms", "x", "z", "event"]].sort_values("t_ms")
        if not ev_df.empty:
            pb.events.extend(
                [{"t_ms": int(t), "x": float(x), "z": float(z), "type": str(ev)} for t, x, z, ev in ev_df.itertuples(index=False, name=None)]
            )

    index_matches: list[dict[str, Any]] = []

    for match_id, match in matches.items():
        # normalize time per match
        all_t: list[int] = []
        for p in match.players.values():
            all_t.extend([t for (t, _, _) in p.positions])
            all_t.extend([int(e["t_ms"]) for e in p.events])

        if not all_t:
            continue

        t0 = min(all_t)
        t1 = max(all_t)
        duration_ms = int(max(0, t1 - t0))

        humans = 0
        bots = 0

        players_out: list[dict[str, Any]] = []
        for p in match.players.values():
            if p.is_bot:
                bots += 1
            else:
                humans += 1

            players_out.append(
                {
                    "userId": p.user_id,
                    "isBot": p.is_bot,
                    "positions": [{"t": int(t - t0), "x": x, "z": z} for (t, x, z) in p.positions],
                    "events": [{"t": int(e["t_ms"] - t0), "x": e["x"], "z": e["z"], "type": e["type"]} for e in p.events],
                }
            )

        match_out = {
            "matchId": match.match_id,
            "mapId": match.map_id,
            "date": match.date,
            "durationMs": duration_ms,
            "players": players_out,
        }

        out_name = f"{match_id}.json"
        out_path = out_matches / out_name
        out_path.write_text(json.dumps(match_out, separators=(",", ":"), ensure_ascii=False), encoding="utf-8")

        index_matches.append(
            {
                "matchId": match.match_id,
                "mapId": match.map_id,
                "date": match.date,
                "durationMs": duration_ms,
                "humans": humans,
                "bots": bots,
                "path": f"/data/{OUT_MATCHES_DIRNAME}/{out_name}",
            }
        )

    index = {
        "maps": list(MAP_CONFIG.keys()),
        "dates": sorted({m["date"] for m in index_matches}),
        "matches": sorted(index_matches, key=lambda m: (m["date"], m["mapId"], m["matchId"])),
        "mapConfig": MAP_CONFIG,
    }

    out_root.mkdir(parents=True, exist_ok=True)
    (out_root / "index.json").write_text(json.dumps(index, indent=2, ensure_ascii=False), encoding="utf-8")

    print(f"Wrote {len(index_matches)} matches to {out_matches}")
    print(f"Wrote index to {out_root / 'index.json'}")


if __name__ == "__main__":
    main()

File: web/scripts/preprocess.mjs
import fs from "node:fs";
import path from "node:path";
import url from "node:url";

import duckdb from "duckdb";

const __dirname = path.dirname(url.fileURLToPath(import.meta.url));

const RAW_DATA_DEFAULT = path.resolve(__dirname, "..", "..", "data", "player_data", "player_data");
const OUT_DATA_DEFAULT = path.resolve(__dirname, "..", "public", "data");
const OUT_MATCHES_DIRNAME = "matches";

const MAP_CONFIG = {
  AmbroseValley: { scale: 900, origin_x: -370, origin_z: -473 },
  GrandRift: { scale: 581, origin_x: -290, origin_z: -290 },
  Lockdown: { scale: 1000, origin_x: -500, origin_z: -500 },
};

function decodeEvent(v) {
  if (Buffer.isBuffer(v)) return v.toString("utf8");
  if (v instanceof Uint8Array) return Buffer.from(v).toString("utf8");
  return String(v);
}

function isBotUserId(userId) {
  return /^\d+$/.test(userId);
}

function isDayFolder(name) {
  return name.startsWith("February_");
}

function ensureDir(p) {
  fs.mkdirSync(p, { recursive: true });
}

function safeReaddir(dir) {
  try {
    return fs.readdirSync(dir, { withFileTypes: true });
  } catch {
    return [];
  }
}

function writeJson(filePath, obj, pretty = false) {
  fs.writeFileSync(filePath, JSON.stringify(obj, null, pretty ? 2 : 0), "utf8");
}

function toMs(ts) {
  // parquetjs-lite returns JS Date for timestamp columns
  if (ts instanceof Date) return ts.getTime();
  if (typeof ts === "number") return ts; // already ms
  return new Date(ts).getTime();
}

function duckdbAll(db, sql) {
  return new Promise((resolve, reject) => {
    db.all(sql, (err, rows) => {
      if (err) reject(err);
      else resolve(rows);
    });
  });
}

function extractDayFromFilename(filename) {
  // .../February_10/<file>
  const m = String(filename).match(/February_\d{2}/);
  return m?.[0] ?? "Unknown";
}

async function main() {
  const rawRoot = path.resolve(process.env.RAW_PLAYER_DATA_DIR ?? RAW_DATA_DEFAULT);
  const outRoot = path.resolve(process.env.OUT_PUBLIC_DATA_DIR ?? OUT_DATA_DEFAULT);
  const outMatches = path.join(outRoot, OUT_MATCHES_DIRNAME);

  if (!fs.existsSync(rawRoot)) {
    console.error(`Raw data folder not found: ${rawRoot}`);
    process.exit(1);
  }

  ensureDir(outMatches);

  // Copy minimaps
  const minimapsSrc = path.join(rawRoot, "minimaps");
  const minimapsDst = path.resolve(outRoot, "..", "minimaps");
  ensureDir(minimapsDst);
  for (const ent of safeReaddir(minimapsSrc)) {
    if (!ent.isFile()) continue;
    const ext = path.extname(ent.name).toLowerCase();
    if (![".png", ".jpg", ".jpeg", ".webp"].includes(ext)) continue;
    fs.copyFileSync(path.join(minimapsSrc, ent.name), path.join(minimapsDst, ent.name));
  }

  const matches = new Map(); // matchId -> { matchId,mapId,date, players: Map<userId, {...}> }

  const db = new duckdb.Database(":memory:");

  // DuckDB can read all parquet files (even without .parquet extension) via glob.
  // filename=true adds a 'filename' column so we can infer the day folder.
  const parquetGlob = path
    .join(rawRoot, "February_*", "*.nakama-0")
    .replaceAll("\\", "/"); // DuckDB prefers forward slashes

  let rows;
  try {
    rows = await duckdbAll(
      db,
      `
      SELECT
        user_id,
        match_id,
        map_id,
        x,
        z,
        ts,
        event,
        filename
      FROM read_parquet('${parquetGlob}', filename=true);
      `,
    );
  } catch (e) {
    console.error("Failed to read parquet via DuckDB. Double-check dataset path:", parquetGlob);
    console.error(e);
    process.exit(1);
  }

  for (const r of rows) {
    const userId = String(r.user_id);
    const matchId = String(r.match_id);
    const mapId = String(r.map_id);
    if (!MAP_CONFIG[mapId]) continue;

    const day = extractDayFromFilename(r.filename);
    const isBot = isBotUserId(userId);

    const match = matches.get(matchId) ?? {
      matchId,
      mapId,
      date: day,
      players: new Map(),
    };
    if (match.mapId !== mapId) continue;
    matches.set(matchId, match);

    const player = match.players.get(userId) ?? {
      userId,
      isBot,
      positions: [],
      events: [],
    };
    match.players.set(userId, player);

    const tMs = toMs(r.ts);
    const x = Number(r.x);
    const z = Number(r.z);
    const event = decodeEvent(r.event);

    if (event === "Position" || event === "BotPosition") {
      player.positions.push({ tMs, x, z });
    } else {
      player.events.push({ tMs, x, z, type: event });
    }
  }

  const indexMatches = [];

  for (const match of matches.values()) {
    let t0 = Number.POSITIVE_INFINITY;
    let t1 = Number.NEGATIVE_INFINITY;

    for (const p of match.players.values()) {
      for (const pt of p.positions) {
        t0 = Math.min(t0, pt.tMs);
        t1 = Math.max(t1, pt.tMs);
      }
      for (const e of p.events) {
        t0 = Math.min(t0, e.tMs);
        t1 = Math.max(t1, e.tMs);
      }
    }
    if (!Number.isFinite(t0) || !Number.isFinite(t1)) continue;

    let humans = 0;
    let bots = 0;

    const playersOut = [];
    for (const p of match.players.values()) {
      if (p.isBot) bots += 1;
      else humans += 1;

      p.positions.sort((a, b) => a.tMs - b.tMs);
      p.events.sort((a, b) => a.tMs - b.tMs);

      playersOut.push({
        userId: p.userId,
        isBot: p.isBot,
        positions: p.positions.map((pt) => ({ t: Math.max(0, pt.tMs - t0), x: pt.x, z: pt.z })),
        events: p.events.map((e) => ({ t: Math.max(0, e.tMs - t0), x: e.x, z: e.z, type: e.type })),
      });
    }

    const durationMs = Math.max(0, t1 - t0);
    const outName = `${match.matchId}.json`;
    writeJson(path.join(outMatches, outName), {
      matchId: match.matchId,
      mapId: match.mapId,
      date: match.date,
      durationMs,
      players: playersOut,
    });

    indexMatches.push({
      matchId: match.matchId,
      mapId: match.mapId,
      date: match.date,
      durationMs,
      humans,
      bots,
      path: `/data/${OUT_MATCHES_DIRNAME}/${outName}`,
    });
  }

  const index = {
    maps: Object.keys(MAP_CONFIG),
    dates: Array.from(new Set(indexMatches.map((m) => m.date))).sort(),
    matches: indexMatches.sort((a, b) => `${a.date}|${a.mapId}|${a.matchId}`.localeCompare(`${b.date}|${b.mapId}|${b.matchId}`)),
    mapConfig: MAP_CONFIG,
  };

  ensureDir(outRoot);
  writeJson(path.join(outRoot, "index.json"), index, true);

  console.log(`Wrote ${indexMatches.length} matches to ${outMatches}`);
  console.log(`Wrote index to ${path.join(outRoot, "index.json")}`);
}

main();

